omegaconf/nodes.py

```diff
@@ -163,6 +164,126 @@ def __deepcopy__(self, memo: Dict[int, Any] = {}) -> "AnyNode":
+        self,
+        value: Any = None,
+        key: Any = None,
+        parent: Optional[Container] = None,
+        is_optional: bool = True,
+        ref_type: Any = Union[Any],
+        element_types: List[Any] = None,  # type: ignore
+    ):
+        self.element_types = element_types
+        super().__init__(
+            parent=parent,
+            value=value,
+            metadata=Metadata(
+                key=key,
+                optional=is_optional,
+                ref_type=ref_type,
+                object_type=type(value),
+            ),
+        )
+
+    def _wrap_node(
+        self,
+        value: Any,
+        key: Any = None,
+        is_optional: bool = True,
+        parent: Optional[BaseContainer] = None,
+        ref_type: Any = None,
+        use_type: bool = False,
+    ) -> Optional[Node]:
+        from omegaconf.omegaconf import _maybe_wrap
+
+        if value is None:
+            return None
+        else:
+            return _maybe_wrap(
+                ref_type=ref_type,
+                key=key,
+                value=value,
+                is_optional=is_optional,
+                parent=parent,
+                use_type=use_type,
+            )
+
+    def validate_and_convert(self, value: Any) -> Optional[Any]:
+        from omegaconf._utils import is_structured_config
+        from omegaconf.basecontainer import BaseContainer
+
+        value_node = None
+        valid_value = False
+        value_type: Any = type(value)
+        if isinstance(value, BaseContainer) and is_structured_config(
+            value._metadata.ref_type
+        ):
+            value_type = value._metadata.ref_type
+            if any(
+                issubclass(value_type, union_type) for union_type in self.element_types
+            ):
+                return value
+            else:
+                self._raise_invalid_value(value, value._metadata.ref_type)
+        element_types = self.element_types
+        if value_type in self.element_types:
+            element_types = self._get_element_types_lead_by(value_type)
+        for union_type in element_types:
+            try:
+                value_node = self._wrap_node(
+                    value=value, ref_type=union_type, use_type=True
+                )
+                valid_value = True
+                break
+            except Exception:
+                pass
+        if not valid_value:
+            self._raise_invalid_value(value, type(value))
+        return value_node
+
+    def _raise_invalid_value(self, value: Any, value_type: Any) -> None:
+        raise ValidationError(
+            "Value '{}' with value_type '{}' is not in '{}'".format(
+                value, value_type, self.element_types
+            )
+        )
+
+    def _get_element_types_lead_by(self, type_: Any) -> List[Any]:
+        element_types = copy.copy(self.element_types)
+        element_types.insert(0, element_types.pop(element_types.index(type_)))
+        return element_types
```

Comment:
I realize that performance is not a major concern, but just in case, it's not too hard to avoid some useless copies:
```suggestion
    def _get_element_types_lead_by(self, type_: Any) -> List[Any]:
        pos = self.element_types.index(type_)
        if pos == 0:
            return self.element_types
        element_types = copy.copy(self.element_types)
        element_types[0], element_types[pos] = element_types[pos], element_types[0]
        return element_types
```